File: internal/payments/models/payment.go
package payments

import (
	"time"

	"github.com/google/uuid"
	"github.com/stripe/stripe-go/v84"
	"gorm.io/datatypes"
	"gorm.io/gorm"
)

type Payment struct {
	ID        uuid.UUID `json:"id" gorm:"type:uuid;primaryKey"`
	CreatedAt time.Time      `json:"createdAt"`
	UpdatedAt time.Time      `json:"updatedAt"`
	DeletedAt gorm.DeletedAt `json:"deletedAt" gorm:"index"`

	AccountID      uint  `json:"accountId" gorm:"index;not null"`
	InvoiceID      *uint `json:"invoiceId" gorm:"index"`
	PlanID         *uint `json:"planId" gorm:"index"`
	SubscriptionID *uint `json:"subscriptionId" gorm:"index"`

	PaymentIntentID string  `json:"paymentIntentId" gorm:"uniqueIndex"`
	ChargeID        *string `json:"chargeId" gorm:"uniqueIndex"`

	Amount      int64  `json:"amount"`
	PlatformFee int64  `json:"platformFee"`
	NetAmount   int64  `json:"netAmount"`
	Currency    string `json:"currency" gorm:"type:varchar(3);not null"`

	Status            stripe.PaymentIntentStatus `json:"status" gorm:"type:varchar(40);not null"`
	PaymentMethodType string                     `json:"paymentMethodType"`

	FailureCode    string `json:"failureCode"`
	FailureMessage string `json:"failureMessage"`

	ProviderMetadata datatypes.JSON `json:"providerMetadata"`
}

type PaymentUpdate struct {
	Status           stripe.PaymentIntentStatus `json:"status"`
	ChargeID         *string                    `json:"chargeId"`
	FailureCode      string                     `json:"failureCode"`
	FailureMessage   string                     `json:"failureMessage"`
	ProviderMetadata datatypes.JSON             `json:"providerMetadata"`
}

File: frontend/apps/web/services/payment.ts
export interface Payment {
  id: string
  createdAt: string
  updatedAt: string
  accountId: number
  invoiceId: number | null
  planId: number | null
  subscriptionId: number | null
  paymentIntentId: string
  chargeId: string | null
  amount: number
  platformFee: number
  netAmount: number
  currency: string
  status: string
  paymentMethodType: string
  failureCode: string
  failureMessage: string
}

export interface PaymentIntentResponse {
  id: string
  client_secret: string
  status: string
  amount: number
  currency: string
  metadata: {
    payment_id: string
    plan_id: string
    user_id: string
  }
}

export interface ConfirmPaymentResponse {
  confirmed: boolean
  paymentId: string
  confirmedUrl: string
}

export function usePaymentService() {
  const { request } = useApiClient()

  return {
    createPaymentIntent: (cardType: string, planId: number) =>
      request<PaymentIntentResponse>('/stripe/paymentIntent', {
        method: 'POST',
        body: JSON.stringify({ cardType, planId }),
      }),

    confirmPaymentIntent: (paymentId: string) =>
      request<ConfirmPaymentResponse>('/stripe/paymentIntent/confirm', {
        method: 'POST',
        body: JSON.stringify({ paymentId }),
      }),

    getPayments: () => request<Payment[]>('/payments'),

    cancelPaymentIntent: () =>
      request<void>('/stripe/paymentIntent/cancel', { method: 'POST' }),
  }
}
